### PR TITLE
Fix permission denied when building symlink derivation which points t…

### PR DESCRIPTION
**Pull request by Artturin**
…o a symlink out of the store

Fix issue #9579

`error: getting attributes of path '/nix/store/ws9yl6ph10v79gx4p9ilhyxg214xf7i7-direct-symlink': Permission denied`

```nix
let
  pkgs = import /home/artturin/nixgits/my-nixpkgs { };
  local = "/home/artturin/nixgits/my-nixpkgs";
in
rec {
  direct-symlink = pkgs.runCommand "direct-symlink" { } ''
    ln -vs ${local}/.version $out
  '';
  indirect-symlink = pkgs.runCommand "indirect-symlink" { } ''
    # Doesn't fail
    # ln -vs ${direct-symlink} $out/not-top
    # Fails
    ln -vs ${direct-symlink} $out
  '';
}
```

# Motivation

# Context

- Meant to address #9579
- Issue introduced by #8965

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libstore/build/local-derivation-goal.cc [hunk 1]
```diff
@@ -398,6 +398,19 @@ static void doBind(const Path & source, const Path & target, bool optional = fal
     if (stat(source.c_str(), &st) == -1) {
         if (optional && errno == ENOENT)
             return;
+        // EACCES can happen when a input path is a symlink to outside the store
+        else if (errno == EACCES) {
+            copyPath(source, target);
+            return;
+            // either link or copy
+            // link doesn't work at all
```
Comment (Artturin) [inline src/libstore/build/local-derivation-goal.cc:406]: Doesn't work in the case of the example in the commit message


Hmm
```
  indirect-symlink = pkgs.runCommand "indirect-symlink" { } ''
    cat ${direct-symlink}
    ln -vs ${direct-symlink} $out
  '';
```
`> cat: /nix/store/k3isri0ks6l63rk54z4714dc67sd1cp3-direct-symlink: No such file or directory`

Comment (Artturin) [reply src/libstore/build/local-derivation-goal.cc:406]: Commented out the auto deletion of the chroot dir

```
$ sudo ls -l /nix/store/q5glq907j6jn39aw99ncmcv5ijvnh5cs-indirect-symlink.drv.chroot/nix/store/55lg749likjz3ivq2varjcbj17xjam40-direct-symlink
lrwxrwxrwx root root 37 B Sat Jan 20 18:13:44 2024  /nix/store/q5glq907j6jn39aw99ncmcv5ijvnh5cs-indirect-symlink.drv.chroot/nix/store/55lg749likjz3ivq2varjcbj17xjam40-direct-symlink@ ⇒ /home/artturin/nixgits/my-nixpkgs/.version
```

I think what is happening is that the symlink to `/home/...` is being mounted/linked/copied to the chroot but the chroot doesn't have access to the result of the symlink

Comment (Artturin) [reply src/libstore/build/local-derivation-goal.cc]: I checked on 2.18.1 and there was still `cat: /nix/store/55lg749likjz3ivq2varjcbj17xjam40-direct-symlink: No such file or directory`, therefore it's not a regression caused by #8965

**Comment by Artturin** [comment]
Tried to make a test but it works in `nix build` and `nix develop` both on master and with this PR

```diff
commit 31554217b0bbf3392125f3c6d24d340d092deec3
Author: Artturin <Artturin@artturin.com>
Date:   2024-01-22 17:57:50 +0200

    Add test for out of store symlink
    
    Issue 9579
    
    "error: getting attributes of path '/nix/store/ws9yl6ph10v79gx4p9ilhyxg214xf7i7-direct-symlink': Permission denied"

diff --git a/tests/functional/local.mk b/tests/functional/local.mk
index 192e275e3..05bebae5d 100644
--- a/tests/functional/local.mk
+++ b/tests/functional/local.mk
@@ -110,6 +110,7 @@ nix_tests = \
   build.sh \
   build-delete.sh \
   output-normalization.sh \
+  out-of-store-symlink.sh \
   selfref-gc.sh \
   db-migration.sh \
   bash-profile.sh \
diff --git a/tests/functional/out-of-store-symlink.nix b/tests/functional/out-of-store-symlink.nix
new file mode 100644
index 000000000..0271b0d9e
--- /dev/null
+++ b/tests/functional/out-of-store-symlink.nix
@@ -0,0 +1,22 @@
+with import ./config.nix;
+
+let
+  direct-symlink = mkDerivation rec {
+    name = "direct-symlink";
+    buildCommand = ''
+      ln -vs ${(builtins.unsafeGetAttrPos "name" direct-symlink).file} $out
+    '';
+    meta.description = "direct-symlink";
+  };
+in
+mkDerivation rec {
+  name = "indirect-symlink";
+  buildCommand = ''
+    # Fails with:
+    # `cat: /nix/store/f2l57qr1lapn3ksyqwivi9848q3g95y4-direct-symlink: No such file or directory`
+    # Probably because the sandbox does not have access to the target of the out of store symlink.
+    # cat ${direct-symlink}
+    ln -vs ${direct-symlink} $out
+  '';
+  meta.description = "direct-symlink";
+}
diff --git a/tests/functional/out-of-store-symlink.sh b/tests/functional/out-of-store-symlink.sh
new file mode 100644
index 000000000..9887966d6
--- /dev/null
+++ b/tests/functional/out-of-store-symlink.sh
@@ -0,0 +1,5 @@
+source common.sh
+
+clearStore
+
+nix build --no-link -f ./out-of-store-symlink.nix
```

**Comment by Ericson2314** [comment]
@Artturin I lost a comment, but this should be using `lstat` not `stat`. The easiest way to do that is actually to use `PosixSourceAccessor` and the methods on it, which will return nicer types (and also be a platform abstraction helper later ("Posix" should be removed from its name)).

**Comment by thufschmitt** [comment]
Discussed during the Nix maintainers meeting on 2024-01-22.
Assigned to @ericson2314

<details>
- @ericson2314: Oops, I thought I left a review here, but now I don't see anything.

- Needs `lstat`

Assigned to @ericson2314

**Comment by Artturin** [comment]
> @Artturin I lost a comment, but this should be using `lstat` not `stat`. The easiest way to do that is actually to use `PosixSourceAccessor` and the methods on it, which will return nicer types (and also be a platform abstraction helper later ("Posix" should be removed from its name)).

If I switch to `lstat` then `error: bind mount from '/nix/store/f2l57qr1lapn3ksyqwivi9848q3g95y4-direct-symlink' to '/nix/store/s1k8hwbh09wnicl9yfpk108vzncjxhl2-indirect-symlink.drv.chroot/nix/store/f2l57qr1lapn3ksyqwivi9848q3g95y4-direct-symlink' failed: Permission denied`

**Comment by Ericson2314** [comment]
@Artturin sorry for missing this. `maybeLstat` has a different return type, that is why the `-1` comparison doesn't work! It returns a `std::optional` instead so you probably want to do

```c++
auto [accessor, canonPath] = PosixSourceAccessor::createAtRoot(tmpFile);
if (auto statOpt = accessor.maybeLstat(canonPath)) {
```

**Comment by alois31** [comment]
> If I switch to `lstat` then `error: bind mount from '/nix/store/f2l57qr1lapn3ksyqwivi9848q3g95y4-direct-symlink' to '/nix/store/s1k8hwbh09wnicl9yfpk108vzncjxhl2-indirect-symlink.drv.chroot/nix/store/f2l57qr1lapn3ksyqwivi9848q3g95y4-direct-symlink' failed: Permission denied`

It is not possible to bind-mount symlinks using `mount`. `S_ISLNK` should be checked explicitly and the source symlink be copied in this case.

**Pull request closed without merging**